File: dsl_planner/src/dstarlite.py
#!/usr/bin/env python3

from nav_core.base_global_planner import BaseGlobalPlanner
from pluginlib import SimplePlugin
import rospy
from nav_msgs.msg import OccupancyGrid, Path
from geometry_msgs.msg import PoseStamped, PoseWithCovarianceStamped
import heapq
import math

class DStarLiteROS(BaseGlobalPlanner, SimplePlugin):
    def __init__(self):
        rospy.init_node('dstar_lite_planner')
        
        # Subscribers
        rospy.Subscriber('/map', OccupancyGrid, self.map_callback)
        rospy.Subscriber('/amcl_pose', PoseWithCovarianceStamped, self.pose_callback)
        rospy.Subscriber('/move_base_simple/goal', PoseStamped, self.goal_callback)
        
        # Publishers
        self.path_pub = rospy.Publisher('/dstar_lite_path', Path, queue_size=1)
        
        # Initialize variables
        self.start = None
        self.goal = None
        self.km = 0
        self.graph = {}
        self.U = []

    def map_callback(self, msg):
        if not self.start or not self.goal:
            # Initialize only once when both start and goal are available
            return
        
        width = msg.info.width
        height = msg.info.height
        data = msg.data

        self.graph = {}
        for i in range(width):
            for j in range(height):
                index = j * width + i
                cost = data[index]
                self.graph[(i, j)] = cost

        self.U = [(self.calculate_key(self.start), self.start)]
        self.compute_shortest_path()

    def pose_callback(self, pose_msg):
        # Update start position based on AMCL pose estimate
        self.start = (int(pose_msg.pose.pose.position.x), int(pose_msg.pose.pose.position.y))
        if self.goal:
            self.map_callback(None)  # Trigger map update

    def goal_callback(self, goal_msg):
        # Update goal position when a new goal is received
        self.goal = (int(goal_msg.pose.position.x), int(goal_msg.pose.position.y))
        self.compute_shortest_path()

    def calculate_key(self, node):
        return (min(self.g(node), self.rhs(node)) + self.heuristic(node, self.start) + self.km, min(self.g(node), self.rhs(node)))

    def heuristic(self, a, b):
        return math.sqrt((a[0] - b[0])**2 + (a[1] - b[1])**2)

    def g(self, node):
        return self.graph.get(node, float('inf'))

    def rhs(self, node):
        if node == self.goal:
            return 0
        return self.graph.get(node, float('inf'))

    def update_vertex(self, node):
        if node != self.start:
            self.graph[node] = min(self.g(node), self.rhs(node))

        if node in self.U:
            self.U.remove(node)
            heapq.heapify(self.U)

        if self.g(node) != self.rhs(node):
            heapq.heappush(self.U, (self.calculate_key(node), node))

    def compute_shortest_path(self):
        if not self.start or not self.goal:
            # Wait until both start and goal are available
            return

        while self.U and (self.U[0][0] < self.calculate_key(self.goal) or self.rhs(self.goal) != self.g(self.goal)):
            k_old, u = heapq.heappop(self.U)

            if k_old < self.calculate_key(u):
                heapq.heappush(self.U, (self.calculate_key(u), u))
            elif self.g(u) > self.rhs(u):
                self.graph[u] = self.rhs(u)
                for succ in self.get_neighbors(u):
                    self.update_vertex(succ)
            else:
                self.graph[u] = float('inf')
                for succ in self.get_neighbors(u) + [u]:
                    self.update_vertex(succ)

        self.publish_path()

    def get_neighbors(self, node):
        x, y = node
        neighbors = []
        for i in range(max(0, x-1), min(width, x+2)):
            for j in range(max(0, y-1), min(height, y+2)):
                if (i, j) != node:
                    neighbors.append((i, j))
        return neighbors

    def publish_path(self):
        path_msg = Path()
        path_msg.header.stamp = rospy.Time.now()
        path_msg.header.frame_id = "map"

        current = self.start
        while current != self.goal:
            pose = PoseStamped()
            pose.pose.position.x = current[0]
            pose.pose.position.y = current[1]
            path_msg.poses.append(pose)

            neighbors = self.get_neighbors(current)
            best_neighbor = min(neighbors, key=lambda n: self.g(n) + self.heuristic(n, self.start))
            current = best_neighbor

        self.path_pub.publish(path_msg)

    def replan(self):
        self.km += self.heuristic(self.start, self.goal)
        self.compute_shortest_path()

if __name__ == '__main__':
    dstar_ros = DStarLiteROS()
    rospy.spin()

File: dsl_planner/src/dstarlite.cpp
#include <ros/ros.h>
#include <nav_msgs/OccupancyGrid.h>
#include <nav_msgs/Path.h>
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/PoseWithCovarianceStamped.h>
#include <queue>
#include <cmath>

class DStarLiteROS {
public:
    DStarLiteROS() {
        ros::NodeHandle nh;
        
        // Subscribers
        map_sub = nh.subscribe("/map", 1, &DStarLiteROS::mapCallback, this);
        pose_sub = nh.subscribe("/amcl_pose", 1, &DStarLiteROS::poseCallback, this);
        goal_sub = nh.subscribe("/move_base_simple/goal", 1, &DStarLiteROS::goalCallback, this);
        
        // Publishers
        path_pub = nh.advertise<nav_msgs::Path>("/dstar_lite_path", 1);
        
        // Initialize variables
        start = std::make_pair(-1, -1);
        goal = std::make_pair(-1, -1);
        km = 0;
    }

    void mapCallback(const nav_msgs::OccupancyGrid::ConstPtr& msg) {
        if (start.first == -1 || start.second == -1 || goal.first == -1 || goal.second == -1) {
            // Initialize only once when both start and goal are available
            return;
        }
        
        int width = msg->info.width;
        int height = msg->info.height;

        graph.clear();
        for (int i = 0; i < width; ++i) {
            for (int j = 0; j < height; ++j) {
                int index = j * width + i;
                int cost = msg->data[index];
                graph[std::make_pair(i, j)] = cost;
            }
        }

        U.push(std::make_pair(calculateKey(start), start));
        computeShortestPath();
    }

    void poseCallback(const geometry_msgs::PoseWithCovarianceStamped::ConstPtr& pose_msg) {
        // Update start position based on AMCL pose estimate
        start.first = static_cast<int>(pose_msg->pose.pose.position.x);
        start.second = static_cast<int>(pose_msg->pose.pose.position.y);

        if (goal.first != -1 && goal.second != -1) {
            mapCallback(nullptr);  // Trigger map update
        }
    }

    void goalCallback(const geometry_msgs::PoseStamped::ConstPtr& goal_msg) {
        // Update goal position when a new goal is received
        goal.first = static_cast<int>(goal_msg->pose.position.x);
        goal.second = static_cast<int>(goal_msg->pose.position.y);
        computeShortestPath();
    }

    double calculateKey(const std::pair<int, int>& node) {
        return (std::min(g(node), rhs(node)) + heuristic(node, start) + km, std::min(g(node), rhs(node)));
    }

    double heuristic(const std::pair<int, int>& a, const std::pair<int, int>& b) {
        return std::sqrt(std::pow(a.first - b.first, 2) + std::pow(a.second - b.second, 2));
    }

    double g(const std::pair<int, int>& node) {
        auto it = graph.find(node);
        return (it != graph.end()) ? it->second : std::numeric_limits<double>::infinity();
    }

    double rhs(const std::pair<int, int>& node) {
        return (node == goal) ? 0 : g(node);
    }

    void updateVertex(const std::pair<int, int>& node) {
        if (node != start) {
            graph[node] = std::min(g(node), rhs(node));
        }

        if (U.count(node) > 0) {
            U.erase(node);
        }

        if (g(node) != rhs(node)) {
            U.insert(std::make_pair(calculateKey(node), node));
        }
    }

    void computeShortestPath() {
        if (start.first == -1 || start.second == -1 || goal.first == -1 || goal.second == -1) {
            // Wait until both start and goal are available
            return;
        }

        while (!U.empty() && (U.begin()->first < calculateKey(goal) || rhs(goal) != g(goal))) {
            auto k_old_u = *U.begin();
            U.erase(U.begin());
            double k_old = k_old_u.first;
            auto u = k_old_u.second;

            if (k_old < calculateKey(u)) {
                U.insert(std::make_pair(calculateKey(u), u));
            } else if (g(u) > rhs(u)) {
                graph[u] = rhs(u);
                for (auto& succ : getNeighbors(u)) {
                    updateVertex(succ);
                }
            } else {
                graph[u] = std::numeric_limits<double>::infinity();
                for (auto& succ : getNeighbors(u)) {
                    updateVertex(succ);
                }
                updateVertex(u);
            }
        }

        publishPath();
    }

    std::vector<std::pair<int, int>> getNeighbors(const std::pair<int, int>& node) {
        int x = node.first;
        int y = node.second;
        std::vector<std::pair<int, int>> neighbors;

        for (int i = std::max(0, x - 1); i < std::min(width, x + 2); ++i) {
            for (int j = std::max(0, y - 1); j < std::min(height, y + 2); ++j) {
                if (std::make_pair(i, j) != node) {
                    neighbors.push_back(std::make_pair(i, j));
                }
            }
        }

        return neighbors;
    }

    void publishPath() {
        nav_msgs::Path path_msg;
        path_msg.header.stamp = ros::Time::now();
        path_msg.header.frame_id = "map";

        auto current = start;
        while (current != goal) {
            geometry_msgs::PoseStamped pose;
            pose.pose.position.x = current.first;
            pose.pose.position.y = current.second;
            path_msg.poses.push_back(pose);

            auto neighbors = getNeighbors(current);
            auto best_neighbor = *std::min_element(neighbors.begin(), neighbors.end(),
                [&](const auto& n1, const auto& n2) {
                    return g(n1) + heuristic(n1, start) < g(n2) + heuristic(n2, start);
                });
            current = best_neighbor;
        }

        path_pub.publish(path_msg);
    }

private:
    ros::Subscriber map_sub;
    ros::Subscriber pose_sub;
    ros::Subscriber goal_sub;
    ros::Publisher path_pub;

    std::pair<int, int> start;
    std::pair<int, int> goal;
    double km;
    std::map<std::pair<int, int>, double> graph;
    std::set<std::pair<double, std::pair<int, int>>> U;

    int width;
    int height;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "dstar_lite_planner_cpp");
    DStarLiteROS dstar_ros;
    ros::spin();
    return 0;
}
